File: app/scripts/controllers/userTable.js
(function(){
	var nmsReportsApp = angular
		.module('nmsReports')
		.controller("UserTableController", ['$scope', '$state', '$filter', 'UserTableFactory', '$localStorage', function($scope, $state, $filter, UserTableFactory, $localStorage) {
            var selectedPage = $scope.currentPageNo;
            var lessRecordsFilter = false;
			$scope.numPerPageList = [10, 20];
            $scope.waiting = UserTableFactory.getAllUsers().length === 0;

			function init() {
				$scope.numPerPage = $scope.numPerPageList[0];
                var filter = $localStorage.filter;
				if (filter) {
					for (key in filter) {
						if (filter[key]) {
							$scope[key] = filter[key];
							if (key === 'filterText') {
								console.log(filter[key]);
								$scope.filterText = filter[key];
							}
						}
					}
				}
                delete $localStorage.filter;
			}

			UserTableFactory.clearAllUsers();
			UserTableFactory.getUsers().then(function(result) {
				UserTableFactory.setUsers(result.data);
				$scope.waiting = false;
                $scope.currentPageNo = selectedPage;
                init();
            });

			function resetPage() {
				$scope.currentPageNo = 1;
			}

			$scope.resetDistrict = function(){
				$scope.districtName = "";
			};

			$scope.resetBlock = function(){
				$scope.blockName = "";
			};

			$scope.getAllUsers = function() {
				var users = UserTableFactory.getAllUsers();
                if ($filter('filter')(users, $scope.search).length <= $scope.numPerPage) {
                    lessRecordsFilter = true;
                    $scope.currentPageNo = 1;
                }
				return users;
			};

			$scope.crop = function(name){
				if(name.length > 13){
					return name.substring(0, 10) + "...";
				}
				return name;
			};

			$scope.getUniqueAccessTypes = function(){
				var aT = [];
				var users = $scope.getAllUsers();
				for(var i=0;i<users.length;i++){
					if(aT.indexOf(users[i].accessType) === -1) {
						aT.push(users[i].accessType);
					}
				}
				return aT;
			};

			$scope.setUniqueAccessTypes = function(accessType) {
                $scope.accType = accessType;
            };

			$scope.getUniqueAccessLevels = function() {
				var aL = [];
				var users = UserTableFactory.getAllUsers();
				for(var i=0;i<users.length;i++){
					if(aL.indexOf(users[i].accessLevel) === -1) {
						aL.push(users[i].accessLevel);
					}
				}
				return aL;
			};

			$scope.setUniqueAccessLevels = function(accessLevel){
				$scope.accLevel = accessLevel;
			};

			$scope.getUniqueStates = function(){
				var states = [];
				var users = UserTableFactory.getAllUsers();
				for(var i=0;i<users.length;i++){
					if((users[i].state !== "") && (states.indexOf(users[i].state) === -1)) {
						states.push(users[i].state);
					}
				}
				return states;
			};
			$scope.setUniqueStates = function(state){
				$scope.stateName = state;
			};
			$scope.$watch('stateName', $scope.resetDistrict);
			$scope.$watch('stateName', $scope.resetBlock);
			$scope.$watch('currentPageNo', function() {
				if (lessRecordsFilter) {
                    $state.transitionTo('userManagement.userTable', {pageNum: $scope.currentPageNo}, {notify: false});
                    lessRecordsFilter = false;
				} else {

				}
            });
			$scope.getUniqueDistricts = function(){
				var districts = [];
				var users = UserTableFactory.getAllUsers();
				for(var i=0;i<users.length;i++){
					if((users[i].state.toLowerCase().indexOf($scope.stateName.toLowerCase() > -1 ||''))&&
					(users[i].block.toLowerCase().indexOf($scope.blockName.toLowerCase() > -1 ||''))&&
					((users[i].district !== "")&&(districts.indexOf(users[i].district) === -1))) {
						districts.push(users[i].district);
					}
				}
				return districts;
			};
			$scope.setUniqueDistricts = function(district){
				$scope.districtName = district;
			};
			$scope.$watch('districtName', $scope.resetBlock);
			$scope.getUniqueBlocks = function(){
				var blocks = [];
				var users = UserTableFactory.getAllUsers();
				for(var i=0;i<users.length;i++){
					if((users[i].state.toLowerCase().indexOf($scope.stateName.toLowerCase()) > -1 ||'') &&
					(users[i].district.toLowerCase().indexOf($scope.districtName.toLowerCase()) > -1 ||'')&&
					((users[i].block !== "") && (blocks.indexOf(users[i].block) === -1))) {
						blocks.push(users[i].block);
					}
				}
				return blocks;
			};
			$scope.setUniqueBlocks = function(block){
				$scope.blockName = block;
			};

			$scope.exists = function(value){
				return value !== '';
			};

			$scope.resetFilters = function(){
				$scope.stateName = '';
				$scope.districtName = '';
				$scope.blockName = '';
				$scope.accType = '';
				$scope.accLevel = '';
				$scope.filterText = '';
				$scope.sorter = 'id';
				$scope.reverse = false;
			};

			$scope.resetFilters();

			$scope.$watch('numPerPage', resetPage());

			$scope.dropdownOpen =function(){
				return (
					$scope.exists($scope.accType) ||
					$scope.exists($scope.accLevel) ||
					$scope.exists($scope.stateName) ||
					$scope.exists($scope.districtName) ||
					$scope.exists($scope.blockName)
				);
			};

			$scope.search = function (row) {
				return (row.name.toLowerCase().indexOf($scope.filterText.toLowerCase() || '') > -1 ||
						row.username.toLowerCase().indexOf($scope.filterText.toLowerCase() || '') > -1 ||
						row.phoneNumber.toLowerCase().indexOf($scope.filterText.toLowerCase()  || '')> -1 ||
						row.email.toLowerCase().indexOf($scope.filterText.toLowerCase() || '')> -1  ||
						row.accessLevel.toLowerCase().indexOf($scope.filterText.toLowerCase()  || '')> -1 ||
						row.state.toLowerCase().indexOf($scope.filterText.toLowerCase()  || '')> -1 ||
						row.district.toLowerCase().indexOf($scope.filterText.toLowerCase()  || '')> -1 ||
						row.block.toLowerCase().indexOf($scope.filterText.toLowerCase() || '') > -1 ) &&(
						row.accessType.toLowerCase().indexOf($scope.accType.toLowerCase()  ||'')> -1 )&&(
						row.accessLevel.toLowerCase().indexOf($scope.accLevel.toLowerCase() ||'') > -1 )&&(
						row.state.toLowerCase().indexOf($scope.stateName.toLowerCase()  ||'') > -1)&&(
						row.district.toLowerCase().indexOf($scope.districtName.toLowerCase()  ||'')> -1 )&&(
						row.block.toLowerCase().indexOf($scope.blockName.toLowerCase()  ||'')> -1
						);
			};

			$scope.sorter = 'id';
			$scope.sort_by = function(newSortingOrder) {
				if ($scope.sorter === newSortingOrder) {
                    $scope.reverse = !$scope.reverse;
                }
				$scope.sorter = newSortingOrder;
			};

			$scope.editUser = function(userId) {
			    setFilterInLocalStorage();
				$state.go("userManagement.editUser", { pageNum: $scope.currentPageNo, id: userId});
			};

			$scope.deleteUser = function(user) {

			};

			$scope.createUser = function() {
                setFilterInLocalStorage();
                $state.go('userManagement.createUser');
			};

			function setFilterInLocalStorage() {
                $localStorage.filter = {
                    filterText: $scope.filterText ? $scope.filterText : null,
                    reverse: $scope.reverse ? $scope.reverse : null,
                    sorter: $scope.sorter ? $scope.sorter : null,
                    accLevel: $scope.accLevel ? $scope.accLevel : null,
                    accType: $scope.accType ? $scope.accType : null,
                    stateName: $scope.stateName ? $scope.stateName : null,
                    districtName: $scope.districtName ? $scope.districtName : null,
                    blockName: $scope.blockName ? $scope.blockName : null
                };
            }

		}])
})();

File: NMSReportingSuite/src/main/java/com/beehyv/nmsreporting/htmlpages/UserTable.java
package com.beehyv.nmsreporting.htmlpages;

public class UserTable {
    public static String pageContent ="<div data-ng-controller=\"UserTableController\"  class='userTable'>\n" +
            "\t<div class=\"block header\">\n" +
            "\t\t<div class='block inputs row'>\n" +
            "\t\t\t<h1 class='col-sm-6'>User Management</h1>\n" +
            "\t\t\t<input type=\"text\" class=\"form-control input input-sm\" data-ng-model=\"filterText\" name=\"tableSearch\" placeholder=\"Search\">\n" +
            "\t\t\t<input type=\"button\" class=\"btn btn-sm btn-primary\" name=\"bulkUpload\" data-ui-sref=\"userManagement.bulkUpload\" value=\"Upload Bulk User\">\n" +
            "\t\t\t<input type=\"button\" class=\"btn btn-sm btn-primary\" name=\"newUser\" data-ng-click=\"createUser()\" value=\"Create new User\">\n" +
            "\t\t</div>\n" +
            "\t</div>\n" +
            "\n" +
            "\t<div>\n" +
            "\t\t<button class=\"btn btn-sm btn-primary resetButton\" data-ng-click=\"resetFilters()\">Reset Filters</button>\n" +
            "\t\t<table class=\"table table-bordered\">\n" +
            "\t\t\t<tr>\n" +
            "\t\t\t\t<th class=\"sortable-columns\" data-ng-click=\"sort_by('id')\"><a>Id</a></th>\n" +
            "\t\t\t\t<th class=\"sortable-columns\" data-ng-click=\"sort_by('name')\"><a>Name</a></th>\n" +
            "\t\t\t\t<th class=\"sortable-columns\" data-ng-click=\"sort_by('username')\"><a>Username</a></th>\n" +
            "\t\t\t\t<th class=\"sortable-columns\" data-ng-click=\"sort_by('phoneNumber')\"><a>Phone no.</a></th>\n" +
            "\t\t\t\t<th class=\"sortable-columns\" data-ng-click=\"sort_by('email')\"><a>Email id</a></th>\n" +
            "\t\t\t\t<th class=\"th-dropdown\">\n" +
            "\t\t\t\t\t<div data-uib-dropdown>\n" +
            "\t\t\t\t\t\t<div data-uib-dropdown-toggle id=\"accessTypeDropdown\">\n" +
            "\t\t\t\t\t\t\t<a href>\n" +
            "\t\t\t\t\t\t\t\tAccess Type\n" +
            "\t\t\t\t\t\t\t\t<span class=\"caret\"></span>\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t\t<a href data-ng-if=\"exists(accType)\">\n" +
            "\t\t\t\t\t\t\t\t({{crop(accType)}})\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t</div>\n" +
            "\t\t\t\t\t\t<ul class=\"dropdown-menu\" data-uib-dropdown-menu aria-labelledby=\"accessTypeDropdown\">\n" +
            "\t\t\t\t\t\t\t<li><a data-ng-click=\"setUniqueAccessTypes('')\">ALL</a></li>\n" +
            "\t\t\t\t\t\t\t<li data-ng-repeat=\"aType in uniqueAT = (getUniqueAccessTypes())\">\n" +
            "\t\t\t\t\t\t\t\t<a data-ng-click=\"setUniqueAccessTypes(aType)\" >{{aType}}</a>\n" +
            "\t\t\t\t\t\t\t</li>\n" +
            "\t\t\t\t\t\t</ul>\n" +
            "\t\t\t\t\t</div>\n" +
            "\t\t\t\t</th>\n" +
            "\t\t\t\t<th class=\"th-dropdown\">\n" +
            "\t\t\t\t\t<div data-uib-dropdown>\n" +
            "\t\t\t\t\t\t<div data-uib-dropdown-toggle id=\"accessLevelDropdown\">\n" +
            "\t\t\t\t\t\t\t<a href>\n" +
            "\t\t\t\t\t\t\t\tAccess Level\n" +
            "\t\t\t\t\t\t\t\t<span class=\"caret\"></span>\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t\t<a href data-ng-if=\"exists(accLevel)\">\n" +
            "\t\t\t\t\t\t\t\t({{crop(accLevel)}})\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t</div>\n" +
            "\t\t\t\t\t\t<ul class=\"dropdown-menu\" data-uib-dropdown-menu aria-labelledby=\"accessLevelDropdown\">\n" +
            "\t\t\t\t\t\t\t<li>\n" +
            "\t\t\t\t\t\t\t\t<a data-ng-click=\"setUniqueAccessLevels('')\">ALL</a>\n" +
            "\t\t\t\t\t\t\t</li>\n" +
            "\t\t\t\t\t\t\t<li data-ng-repeat=\"aLevel in uniqueAL = (getUniqueAccessLevels())\">\n" +
            "\t\t\t\t\t\t\t\t<a data-ng-click=\"setUniqueAccessLevels(aLevel)\" >{{aLevel}}</a>\n" +
            "\t\t\t\t\t\t\t</li>\n" +
            "\t\t\t\t\t\t</ul>\n" +
            "\t\t\t\t\t</div>\n" +
            "\t\t\t\t</th>\n" +
            "\t\t\t\t<th class=\"th-dropdown\">\n" +
            "\t\t\t\t\t<div data-uib-dropdown>\n" +
            "\t\t\t\t\t\t<div data-uib-dropdown-toggle id=\"stateDropdown\">\n" +
            "\t\t\t\t\t\t\t<a href>\n" +
            "\t\t\t\t\t\t\t\tState\n" +
            "\t\t\t\t\t\t\t\t<span class=\"caret\"></span>\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t\t<a href data-ng-if=\"exists(stateName)\">\n" +
            "\t\t\t\t\t\t\t\t({{crop(stateName)}})\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t</div>\n" +
            "\t\t\t\t\t\t<ul class=\"dropdown-menu\" data-uib-dropdown-menu aria-labelledby=\"stateDropdown\">\n" +
            "\t\t\t\t\t\t\t<li>\n" +
            "\t\t\t\t\t\t\t\t<a data-ng-click=\"setUniqueStates('')\">ALL</a>\n" +
            "\t\t\t\t\t\t\t</li>\n" +
            "\t\t\t\t\t\t\t<li data-ng-repeat=\"sName in sNames = (getUniqueStates()) | orderBy\">\n" +
            "\t\t\t\t\t\t\t\t<a data-ng-click=\"setUniqueStates(sName)\" >{{sName}}</a>\n" +
            "\t\t\t\t\t\t\t</li>\n" +
            "\t\t\t\t\t\t</ul>\n" +
            "\t\t\t\t\t</div>\n" +
            "\t\t\t\t</th>\n" +
            "\t\t\t\t<th class=\"th-dropdown\">\n" +
            "\t\t\t\t\t<div data-uib-dropdown>\n" +
            "\t\t\t\t\t\t<div data-uib-dropdown-toggle id=\"districtDropdown\">\n" +
            "\t\t\t\t\t\t\t<a href>\n" +
            "\t\t\t\t\t\t\t\tDistrict\n" +
            "\t\t\t\t\t\t\t\t<span class=\"caret\"></span>\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t\t<a href data-ng-if=\"exists(districtName)\">({{crop(districtName)}})\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t</div>\n" +
            "\t\t\t\t\t\t<ul class=\"dropdown-menu\" data-uib-dropdown-menu aria-labelledby=\"districtDropdown\">\n" +
            "\t\t\t\t\t\t\t<li>\n" +
            "\t\t\t\t\t\t\t\t<a data-ng-click=\"setUniqueDistricts('')\">ALL</a>\n" +
            "\t\t\t\t\t\t\t</li>\n" +
            "\t\t\t\t\t\t\t<li data-ng-repeat=\"dName in dNames = (getUniqueDistricts()) | orderBy\">\n" +
            "\t\t\t\t\t\t\t\t<a data-ng-click=\"setUniqueDistricts(dName)\" >{{dName}}</a>\n" +
            "\t\t\t\t\t\t\t</li>\n" +
            "\t\t\t\t\t\t</ul>\n" +
            "\t\t\t\t\t</div>\n" +
            "\t\t\t\t</th>\n" +
            "\t\t\t\t<th class=\"th-dropdown\">\n" +
            "\t\t\t\t\t<div data-uib-dropdown>\n" +
            "\t\t\t\t\t\t<div data-uib-dropdown-toggle id=\"blockDropdown\">\n" +
            "\t\t\t\t\t\t\t<a href>Block\n" +
            "\t\t\t\t\t\t\t\t<span class=\"caret\"></span>\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t\t<a href data-ng-if=\"exists(blockName)\">\n" +
            "\t\t\t\t\t\t\t\t({{crop(blockName)}})\n" +
            "\t\t\t\t\t\t\t</a>\n" +
            "\t\t\t\t\t\t</div>\n" +
            "\t\t\t\t\t\t<ul class=\"dropdown-menu\" data-uib-dropdown-menu aria-labelledby=\"blockDropdown\">\n" +
            "\t\t\t\t\t\t\t<li><a data-ng-click=\"setUniqueBlocks('')\">ALL</a></li>\n" +
            "\t\t\t\t\t\t\t<li data-ng-repeat=\"bName in bNames = (getUniqueBlocks()) | orderBy\">\n" +
            "\t\t\t\t\t\t\t\t<a data-ng-click=\"setUniqueBlocks(bName)\" >{{bName}}</a>\n" +
            "\t\t\t\t\t\t\t</li>\n" +
            "\t\t\t\t\t\t</ul>\n" +
            "\t\t\t\t\t</div>\n" +
            "\t\t\t\t</th>\n" +
            "\t\t\t\t<th><a>Edit</a></th>\n" +
            "\t\t\t</tr>\n" +
            "\t\t\t<tr data-ng-repeat=\"item in filterData = (getAllUsers() | filter : search ) | orderBy : sorter : reverse  | limitTo : numPerPage : numPerPage * (currentPageNo - 1)\">\n" +
            "\t\t\t\t<td>{{item.id - 1}}</td>\n" +
            "\t\t\t\t<td>{{item.name}}</td>\n" +
            "\t\t\t\t<td>{{item.username}}</td>\n" +
            "\t\t\t\t<td>{{item.phoneNumber}}</td>\n" +
            "\t\t\t\t<td>{{item.email}}</td>\n" +
            "\t\t\t\t<td>{{item.accessType}}</td>\n" +
            "\t\t\t\t<td>{{item.accessLevel}}</td>\n" +
            "\t\t\t\t<td>{{item.state}}</td>\n" +
            "\t\t\t\t<td>{{item.district}}</td>\n" +
            "\t\t\t\t<td>{{item.block}}</td>\n" +
            "\t\t\t\t<td><span data-ng-if=\"item.createdBy\"><a class=\"pointer\" data-ng-click=\"editUser(item.id)\">Edit</a></span></td>\n" +
            "\t\t\t</tr>\n" +
            "\t\t</table>\n" +
            "\n" +
            "\t\t<div data-ng-if=\"waiting\" class=\"loading\">\n" +
            "\t\t\t<img src=\"images/pageloader.gif\" alt=\"loading...\">\n" +
            "\t\t</div>\n" +
            "\n" +
            "\t\t<div class=\"block pagination\">\n" +
            "\t\t\t<pagination  total-items=\"filterData.length\" items-per-page=\"numPerPage\"\n" +
            "\t\t\t\t\t\t current-page=\"currentPageNo\" class=\"pagination-small\">\n" +
            "\t\t\t</pagination>\n" +
            "\t\t</div>\n" +
            "\t</div>" +
            "</div>";
}
